test/functionalTest/cases/4085_custom_notifications_ngsi_payload/custom_notification_http_ngsi_adding_full_replacement_attribute_simple.test

```diff
@@ -0,0 +1,392 @@
+===================================================================
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: /v2/subscriptions/REGEX([0-9a-f]{24})
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+
+
+02. Create entity E1
+====================
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: /v2/entities/E1?type=T
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+
+
+03. Update entity E1
+====================
+HTTP/1.1 204 No Content
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+
+
+04. Dump accumulator and see two notifications with replaced values
+===================================================================
+POST http://127.0.0.1:REGEX(\d+)/notify
+Fiware-Servicepath: /
+Content-Length: 731
+User-Agent: orion/REGEX(\d+\.\d+\.\d+.*)
+Ngsiv2-Attrsformat: custom
+Host: 127.0.0.1:REGEX(\d+)
+Accept: application/json
+Content-Type: application/json; charset=utf-8
+Fiware-Correlator: REGEX([0-9a-f\-]{36}); cbnotif=1
+
+{
+    "data": [
+        {
+            "A1": {
+                "metadata": {},
+                "type": "Text",
+                "value": "foo"
+            },
+            "A2": {
+                "metadata": {},
+                "type": "Number",
+                "value": 1
+            },
+            "A3": {
+                "metadata": {},
+                "type": "Boolean",
+                "value": true
+            },
+            "A4": {
+                "metadata": {},
+                "type": "Object",
+                "value": {
+                    "x": 1,
+                    "y": 2
+                }
+            },
+            "A5": {
+                "metadata": {},
+                "type": "Array",
+                "value": [
+                    1,
+                    true,
+                    "zzz"
+                ]
+            },
+            "A6": {
+                "metadata": {},
+                "type": "None",
+                "value": null
+            },
+            "B1": {
+                "metadata": {},
+                "type": "irrelevant",
+                "value": "foo"
+            },
+            "B2": {
+                "metadata": {},
+                "type": "irrelevant",
+                "value": 1
+            },
+            "B3": {
+                "metadata": {},
+                "type": "irrelevant",
+                "value": true
+            },
+            "B4": {
+                "metadata": {},
+                "type": "irrelevant",
+                "value": {
+                    "x": 1,
+                    "y": 2
+                }
+            },
+            "B5": {
+                "metadata": {},
+                "type": "irrelevant",
+                "value": [
+                    1,
+                    true,
+                    "zzz"
+                ]
+            },
+            "B6": {
+                "metadata": {},
+                "type": "irrelevant",
+                "value": null
```

Comment:
Note the "source" attributes (B1 to B6) will appear in the notification. If we avoid them using `notification.attrs` set to `["A1", "A2", "A3", "A4", "A5", "A6" ]` then no notification will be sent due to the old "notification empty-ness checking" at processOnChangeConditionForUpdateContext() in MongoCommonUpdate.cpp:

```
  /* Early exit without sending notification if attribute list is empty */
  if (cer.entity.attributeVector.size() == 0)
  {
    ncr.contextElementResponseVector.release();
    return false;
  }
```

However, even in the case we use some attribute to "trigger" notifications, e.g. `notification.attrs` set to `["A1", "A2", "A3", "A4", "A5", "A6", "B6" ]` will not work as expected. The notifications will be sent, but only B6 replacement will be done. It seems that with the current implementation the attribute has to be part of the notification in order being a "source" for replacements. This is probably due to this code in processOnChangeConditionForUpdateContext() in MongoCommonUpdate.cpp

```
  for (unsigned int ix = 0; ix < notifyCerP->entity.attributeVector.size(); ix++)
  {
    ...
    if ((attrL.size() == 0) || attrL.lookup(ALL_ATTRS) || (blacklist == true))
    {
    ...
    }
    else
    {
      for (unsigned int jx = 0; jx < attrL.size(); jx++)
      {
      ...
      }
    }
  }
```

which passes attributes from `notifyCerP` (the entity in memory, will full content) and `cer.entity.attributeVector` (the one used in the notification).

We should check the exact requirement here, if we can live with this limitation or, in negative case, fix the implementation so we can overcome the limitation but without breaking backward compatibility.

Reply:
Reviewing this with @mrutid we have concluded that it would be great to properly support replacements without needed to have the "source" attributes in the final notification but that it's not a strong requirement (in other words, if the modification on the code to get it is to hard, we can live with this limitation).

Reply:
Notification logic changes (basically, move that filtering logic later in the process of rendering the notification) done in [97f7733](https://github.com/telefonicaid/fiware-orion/pull/4229/commits/97f7733a72e4a1e2f575313618499cb9039c8894) tu support that kind of cases.